File: 322-coin-change/coin-change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        
        //sort coins first
        Arrays.sort(coins);

        //dp[] = minimum number of coins to make an amount
        int[] dp = new int[amount + 1];

        //fill dp array with a theoretical infinite (amount will always take less than amount coins)
        Arrays.fill(dp, amount + 1);

        //0 coins to make amount 0
        dp[0] = 0;

        //loop through each amount and for each amount loop through each denomination
        for(int i = 0; i <= amount; i++){
            for(int j = 0; j < coins.length; j++){

                //if denomination is lower than amount
                if(coins[j] <= i){

                    //assign min coins to min(previous val, coin + val[amount - caoin]) 
                    dp[i] = Math.min(dp[i], 1 + dp[i - coins[j]]);

                }
                else{
                    break;
                }

            }
        }

        //return final ans if lower than amount
        return dp[amount] > amount ? -1 : dp[amount];
    }
}

File: 322-coin-change/coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        int n = coins.size();
        // dp[i][amt] = min coins to form amt using coins[i..]
        vector<vector<int>> dp(n, vector<int>(amount + 1, -1));
        
        int ans = dfs(coins, amount, 0, dp);
        return (ans >= 1e9 ? -1 : ans);
    }

    int dfs(vector<int>& coins, int amount, int i, vector<vector<int>>& dp) {
        if (amount == 0) return 0;                 // no coins needed
        if (i >= coins.size() || amount < 0) return 1e9; // impossible case

        if (dp[i][amount] != -1) return dp[i][amount];

        // take this coin (stay at same index because we can reuse unlimited times)
        int take = 1 + dfs(coins, amount - coins[i], i, dp);

        // skip this coin (move to next index)
        int skip = dfs(coins, amount, i + 1, dp);

        return dp[i][amount] = min(take, skip);
    }
};
